core/commands/files/files.go

```diff
@@ -428,10 +435,328 @@ Examples:
+
+				if item.Depth > 0 && item.LastChild {
+					fmt.Fprint(w, "└── ")
+				}
+
+				if item.Depth == 0 {
+					fmt.Fprintln(w, output.Hash)
+				} else {
+					fmt.Fprintln(w, item.Name)
+				}
+
+				return nil
+			}
+
+			if output.Type == summaryType {
+				summary := output.Summary
+
+				fmt.Fprintf(w, "%d directories, %d files present\n",
+					summary.NbDir,
+					summary.NbFile,
+				)
+
+				if summary.SizeTotal > 0 {
+					fmt.Fprintf(w, "%s present of %s (%.2f%%)",
+						humanize.Bytes(uint64(summary.SizeLocal)),
+						humanize.Bytes(uint64(summary.SizeTotal)),
+						100.0*float64(summary.SizeLocal)/float64(summary.SizeTotal),
+					)
+				} else {
+					fmt.Fprintf(w, "%s present of unknown", humanize.Bytes(uint64(summary.SizeLocal)))
+				}
+
+				fmt.Fprintln(w)
+				return nil
+			}
+
+			return fmt.Errorf("unknown output type")
+		}),
+	},
+}
+
+type walkState struct {
+	ctx     context.Context
+	res     cmds.ResponseEmitter
+	dagserv dag.DAGService
+	node    node.Node
+
+	onlySummary       bool
+	depth             int
+	name              string
+	parentIsDirectory bool
+	lastChild         bool
+}
+
+func walkBlockStart(ctx context.Context, res cmds.ResponseEmitter, dagserv dag.DAGService, nd node.Node, isDirectory bool, onlySummary bool) (treeSummary, error) {
+	return walkBlock(walkState{
+		ctx,
+		res,
+		dagserv,
+		nd,
+		onlySummary,
+		0,
+		nd.Cid().String(),
+		isDirectory,
+		false,
+	})
+}
+
+// Build a summary of a dag while emiting treeOutput with treeItem to display the graph of files/directories
+func walkBlock(state walkState) (treeSummary, error) {
+
+	// Start with the block data size
+	result := treeSummary{
+		Local:     true,
+		SizeLocal: uint64(len(state.node.RawData())),
+	}
+
+	isDirectory := false
+	// try to decode unixfs
+	if pn, ok := state.node.(*dag.ProtoNode); ok {
+		if unixfs, err := ft.FromBytes(pn.Data()); err == nil {
+
+			// if cumulative size is available, use it as total size
+			cumSize, err := pn.Size()
+			if err == nil {
+				result.SizeTotal = cumSize
+			}
+
+			unixType := unixfs.GetType()
+
+			// To distinguish files from chunk
+			isFile := unixType == ft.TFile && state.parentIsDirectory
+			isDirectory = unixType == ft.TDirectory
+
+			if isFile {
+				result.NbFile++
+			}
+
+			// Don't count the root directory if any
+			if isDirectory && state.depth != 0 {
+				result.NbDir++
+			}
+
+			if !state.onlySummary && (isFile || isDirectory) {
+				state.res.Emit(&treeOutput{
+					Hash: state.node.Cid().String(),
+					Type: itemType,
+					Item: treeItem{
+						Name:      state.name,
+						Depth:     state.depth,
+						LastChild: state.lastChild,
+					},
+				})
+			}
+		}
+	}
+
+	nbChild := len(state.node.Links())
+	for i, link := range state.node.Links() {
+		child, err := state.dagserv.Get(state.ctx, link.Cid)
+
+		isLastChild := i == nbChild-1
+
+		if err == dag.ErrNotFound {
+			result.Local = false
+
+			if !state.onlySummary {
+				state.res.Emit(&treeOutput{
```

Comment:
This will make missing blocks in a file look like missing files, if we do display them, I'd just display them as a list of missing blocks (no depth, although you can still pass the depth through if you want).

We should also *explicitly* mark them as missing blocks. As a matter of fact, we should probably include the types of directories and files as well. That is, I'd expect these "events" to be of the form:

```
{
  "type": "directory|file|missing",
  "size": 200, // recursive
  "depth": 1,
  "lastChild": true|false,
  "hash": "QmId",
  "name": "thing" // only applies to named things (files and directories, not missing blocks).
}
```

Note the lack of summary etc, we can compute that in PostRun.

Reply:
I would separate the "type" from the status (missing or non-missing).

Even if a file is missing we still know some information about it from the directory info.  When the new unixfs lands we should also know its exact type.

I would use `directory|file|block|unknown` for the possible types.  `unknown` is used when we have a directory entry but we don't know if it is a file or directory.  `block` is used when we know its a part of a file.

Reply:
I would collapse it, and instead of using the `missing-block` use
```
{
  "type": "unknown",
  "size": 400,
  "depth": 2,
  "hash": "QmSomething",
  "name": "file-or-directory"
  "lastChild": true,
  "status": "missing"
}
```
or something close to that.

Reply:
My issue with that is that we might have missing child-blocks and I'd rather not have two different ways to indicate that we have missing information.

On the other hand, missing the top-level block (currently) means that we don't even know if we're dealing with a file or a directory so I guess it is a bit of a special case (at the moment).

Reply:
if my missing child block you mean part of a file then I would use:
```
{
  "type": "block",
  "size": 400,
  "depth": 2,
  "hash": "QmSomething",
  "status": "missing"
}
```
I am not sure what the purpose of the `lastChild` field is.

Reply:
> if my missing child block you mean part of a file then I would use...

I understand, I'm just trying to reduce the number of cases we have to handle (although I don't fell strongly about this). This is easier to see when assuming that we already encoded filetype information in directory entries:

Missing all blocks (including the root the root):

```
{
  "type": "file",
  "size": 400,
  "depth": 2,
  "hash": "QmFile",
  "name": "myFile"
  "status": "missing"
}
```

Missing a child block:

```
{
  "type": "file",
  "size": 400,
  "depth": 2,
  "hash": "QmFileRoot",
  "name": "myFile"
}
{
  "type": "block",
  "size": 400,
  "depth": 3,
  "hash": "QmFileBlock1",
  "name": "file-or-directory",
  "status": "missing",
  "lastChild": true
}
```

From a filesystem perspective, both cases are missing (at least one) block of the file. However, we end up representing this differently in both cases.

On the other hand, this is nice because the first case makes it clear that we're missing the *entire* file.

> I am not sure what the purpose of the lastChild field is.

Sorry, that wasn't really necessary for that example. `lastChild` means the last child at that depth. It allows us to draw the tree without buffering (the current implementation has it for this reason).

Reply:
> On the other hand, this is nice because the first case makes it clear that we're missing the entire file.

And I think that is an important distinction.  For the case when some of the children are missing I would use `"status": "incomplete"`.

(I have thought through many of these case when working with the original filestore, that never made it into ipfs, `verify` commands.  You can find the documentation to that here https://github.com/ipfs-filestore/go-ipfs/blob/master/core/commands/filestore.go#L547.  Many of the possible codes don't make since but the 'ok', 'missing' and 'incomplete' codes do.)

Since the original goal of the p.r. was to include information on how much of a tree is downloaded I would also include a field that is the cumulative size found locally (although not sure what to call it).  This will work nicely to avoid having to include summary information in the API as the important information will be available by just looking at the root node.

Reply:
> And I think that is an important distinction. For the case when some of the children are missing I would use `"status": "incomplete"`.

My point is that thinking of some blocks as "children" and others not is kind of pointless. From the perspective of the filesystem, a file has an array of blocks (no parents and children).

> Since the original goal of the p.r. was to include information on how much of a tree is downloaded I would also include a field that is the cumulative size found locally (although not sure what to call it). This will work nicely to avoid having to include summary information in the API as the important information will be available by just looking at the root node.

Unfortunately, we can't compute that until we've processed all the leaves (well after we've sent this event).